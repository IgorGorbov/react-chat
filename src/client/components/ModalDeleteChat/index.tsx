import * as React from 'react';
import { Modal, Button } from 'semantic-ui-react';
import { UserButton } from '../../styledComponents/index';

interface IModalDeleteChatProps {
  userName: string;
  currentChatId: number;
  deleteChat: (chatId: number) => { type: string; payload: number };
}

interface IModalDeleteChatState {
  modalOpen: boolean;
}

class ModalDeleteChat extends React.Component<
  IModalDeleteChatProps,
  IModalDeleteChatState
> {
  public state = { modalOpen: false };
  public handleOpen = (): void => this.setState({ modalOpen: true });
  public handleClose = (): void => {
    this.setState({ modalOpen: false });
  };

  public handleClick(chatId: number): void {
    const { deleteChat } = this.props;
    deleteChat(chatId);
    this.setState({ modalOpen: false });
  }
  public render() {
    const { currentChatId, userName } = this.props;
    const DeleteButton: any = (
      <UserButton
        buttonwidth="40px"
        buttonmargin="0 10px 0 auto"
        buttonheight="35px"
        color="red"
        icon="user delete"
        onClick={(): void => this.handleOpen()}
      />
    );
    return (
      <Modal
        trigger={DeleteButton}
        open={this.state.modalOpen}
        onClose={this.handleClose}
      >
        <Modal.Header>Delete chat</Modal.Header>
        <Modal.Content>
          <p>Are you sure you want to delete chat with {userName}?</p>
        </Modal.Content>
        <Modal.Actions>
          <Button onClick={(): void => this.handleClose()} negative={true}>
            No
          </Button>
          <Button
            onClick={(): void => this.handleClick(currentChatId)}
            positive={true}
            labelPosition="right"
            icon="checkmark"
            content="Yes"
          />
        </Modal.Actions>
      </Modal>
    );
  }
}

export default ModalDeleteChat;
