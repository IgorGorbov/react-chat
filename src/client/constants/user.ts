export enum Status {
  Online = 'Online',
  Offline = 'Offline',
  Away = 'Away',
  Busy = 'Busy',
}

export const statusColors = {
  [Status.Online]: 'green',
  [Status.Offline]: 'grey',
  [Status.Away]: '#dede0f',
  [Status.Busy]: 'red',
};
