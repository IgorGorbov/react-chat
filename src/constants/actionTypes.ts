export const USER_LOGIN = 'USER_LOGIN';
export const CHANGE_USER_STATUS = 'CHANGE_USER_STATUS';
export const FETCH_ALL_USERS = 'FETCH_ALL_USERS';
export const SEARCH_CHAT = 'SEARCH_CHAT';
export const CHANGE_CHAT = 'CHANGE_CHAT';
export const NEW_CHAT = 'NEW_CHAT';
export const DELETE_CHAT = 'DELETE_CHAT';
export const NEW_MESSAGE = 'NEW_MESSAGE';
export const CHOOSE_USER = 'CHOOSE_USER';
export const SELECT_MESSAGE = 'SELECT_MESSAGE';
export const DELETE_MESSAGES = 'DELETE_MESSAGES';
export const READ_MESSAGES = 'READ_MESSAGES';
