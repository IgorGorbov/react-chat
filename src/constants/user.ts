export const defaultAvatar = 'https://emilcarlsson.se/assets/mikeross.png';
export enum Status {
  Online = "Online",
  Offline = "Offline",
  Away = "Away",
  Busy = "Busy"
}

export const statusColors = {
  [Status.Online]: 'green',
  [Status.Offline]: 'grey',
  [Status.Away]: 'yellow',
  [Status.Busy]: 'red'
};
