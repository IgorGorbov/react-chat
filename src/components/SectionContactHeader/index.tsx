import * as React from 'react';
import ModalDeleteChat from '../../components/ModalDeleteChat';
import { statusColors } from '../../constants/user';
import {
  SectionContactHeader as ContactHeader,
  ContactInfoContainer,
  UserButton,
  UserAvatar,
  Text,
} from '../../styledComponents';

interface InterfaceContactHeaderProps {
  currentUser: IUser;
  user: IUser;
  currentChatId: number;
  selectedMessages: number[];
  deleteChat: (chatId: number) => { type: string; payload: number };
  deleteMessages: (
    selectedMessages: number[],
    currentUserID: number,
    currentChatId: number,
  ) => {
    type: string;
    payload: {
      selectedMessages: number[];
      currentUserID: number;
      currentChatId: number;
    };
  };
}

const SectionContactHeader = ({
  currentUser,
  user,
  currentChatId,
  deleteChat,
  selectedMessages,
  deleteMessages,
}: InterfaceContactHeaderProps) => {
  const isSelectedMessages: boolean = selectedMessages.length > 0;
  return (
    <ContactHeader>
      <UserAvatar
        src={user.avatar}
        size="tiny"
        circular={true}
        maxheight="65px"
        maxwidth="65px"
        bordersize="2px"
        bordercolor="grey"
        cursor="false"
      />
      <ContactInfoContainer>
        <Text
          textColor="#000"
          textSize="20px"
          textLineHeight="1.2"
          textWeight="500"
          textFontFamily="Titillium Web"
        >
          {user.name}
        </Text>
        <Text
          textColor={statusColors[user.status]}
          textSize="15px"
          textLineHeight="1"
          textWeight="500"
          textFontFamily="Titillium Web"
        >
          {user.status}
        </Text>
      </ContactInfoContainer>
      {isSelectedMessages ? (
        <UserButton
          buttonwidth="40px"
          buttonmargin="0 10px 0 auto"
          buttonheight="35px"
          color="red"
          icon="trash alternate outline"
          onClick={() =>
            deleteMessages(selectedMessages, currentUser.id, currentChatId)
          }
        />
      ) : (
        <ModalDeleteChat
          userName={user.name}
          currentChatId={currentChatId}
          deleteChat={deleteChat}
        />
      )}
    </ContactHeader>
  );
};

export default SectionContactHeader;
