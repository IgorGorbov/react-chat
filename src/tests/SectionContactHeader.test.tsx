import * as React from 'react';
import toJSON from 'enzyme-to-json';
import { mount } from 'enzyme';

import SectionContactHeader from '../client/components/SectionContactHeader';
import ModalDeleteChat from '../client/components/ModalDeleteChat';
import {
  SectionContactHeader as ContactHeader,
  UserButton,
  UserAvatar,
  ContactInfoContainer,
  Text,
} from '../client/styledComponents/index';

import { Status } from '../client/constants/user';

describe('testing SectionContactHeader', () => {
  interface InterfaceContactHeaderProps {
    currentUser: IUser;
    user: IUser;
    currentChatId: number;
    selectedMessages: number[];
    deleteChat: (chatId: number) => { type: string; payload: number };
    deleteMessages: (
      selectedMessages: number[],
      currentUserID: number,
      currentChatId: number,
    ) => {
      type: string;
      payload: {
        selectedMessages: number[];
        currentUserID: number;
        currentChatId: number;
      };
    };
  }
  const deleteMessagesHandle = jest.fn();
  const deleteChatHandle = jest.fn();

  const props: InterfaceContactHeaderProps = {
    currentUser: {
      id: 1,
      name: 'test',
      avatar: 'test',
      status: Status.Online,
    },
    user: {
      id: 2,
      name: 'test2',
      avatar: 'test2',
      status: Status.Online,
    },
    currentChatId: 1,
    selectedMessages: [],
    deleteChat: deleteChatHandle,
    deleteMessages: deleteMessagesHandle,
  };

  let wrapper: any;
  beforeAll(() => {
    wrapper = mount(<SectionContactHeader {...props} />);
  });

  it('should contain ContactHeader', () => {
    expect(wrapper.find(ContactHeader)).toHaveLength(1);
  });

  it('should contain UserAvatar', () => {
    expect(wrapper.find(UserAvatar)).toHaveLength(1);
  });

  it('should contain ContactInfoContainer', () => {
    expect(wrapper.find(ContactInfoContainer)).toHaveLength(1);
  });

  it('should contain Button(ModalDeleteChat)', () => {
    expect(wrapper.find(ModalDeleteChat)).toHaveLength(1);
  });

  it('should contain user name and status', () => {
    const container = wrapper.find(ContactInfoContainer);
    expect(container.find(Text)).toHaveLength(2);
    expect(
      container
        .find(Text)
        .first()
        .text(),
    ).toEqual(props.user.name);
    expect(
      container
        .find(Text)
        .last()
        .text(),
    ).toEqual(props.user.status);
  });

  it('should contain UserButton(delete messages)', () => {
    const propsSelectedMessages = {
      currentUser: {
        id: 1,
        name: 'test',
        avatar: 'test',
        status: Status.Online,
      },
      user: {
        id: 2,
        name: 'test2',
        avatar: 'test2',
        status: Status.Online,
      },
      currentChatId: 1,
      selectedMessages: [1, 2],
      deleteChat: jest.fn(),
      deleteMessages: deleteMessagesHandle,
    };
    const wrapper = mount(<SectionContactHeader {...propsSelectedMessages} />);
    const ub = wrapper.find(UserButton);
    expect(ub).toHaveLength(1);

    ub.simulate('click');
    expect(deleteMessagesHandle).toHaveBeenCalledTimes(1);
  });

  it('snapshot CurrentChatContainer', () => {
    const wrapper = mount(<SectionContactHeader {...props} />);
    const tree = toJSON(wrapper);
    expect(tree).toMatchSnapshot();
  });
});
