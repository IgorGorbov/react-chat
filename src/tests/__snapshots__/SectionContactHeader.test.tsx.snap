// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing SectionContactHeader snapshot CurrentChatContainer 1`] = `
<SectionContactHeader
  currentChatId={1}
  currentUser={
    Object {
      "avatar": "test",
      "id": 1,
      "name": "test",
      "status": "Online",
    }
  }
  deleteChat={[MockFunction]}
  deleteMessages={
    [MockFunction] {
      "calls": Array [
        Array [
          Array [
            1,
            2,
          ],
          1,
          1,
        ],
      ],
    }
  }
  selectedMessages={Array []}
  user={
    Object {
      "avatar": "test2",
      "id": 2,
      "name": "test2",
      "status": "Online",
    }
  }
>
  <styled.section>
    <section
      className="sc-gzVnrw dwhOpH"
    >
      <Styled(Image)
        bordercolor="grey"
        bordersize="2px"
        circular={true}
        cursor="false"
        maxheight="65px"
        maxwidth="65px"
        size="tiny"
        src="test2"
      >
        <Image
          as="img"
          bordercolor="grey"
          bordersize="2px"
          circular={true}
          className="sc-ifAKCX iWdRSQ"
          cursor="false"
          maxheight="65px"
          maxwidth="65px"
          size="tiny"
          src="test2"
          ui={true}
        >
          <img
            bordercolor="grey"
            bordersize="2px"
            className="ui tiny circular image sc-ifAKCX iWdRSQ"
            cursor="false"
            maxheight="65px"
            maxwidth="65px"
            src="test2"
          />
        </Image>
      </Styled(Image)>
      <styled.div>
        <div
          className="sc-htoDjs dXmIuO"
        >
          <styled.span
            textColor="#000"
            textFontFamily="Titillium Web"
            textLineHeight="1.2"
            textSize="20px"
            textWeight="500"
          >
            <span
              className="sc-EHOje icjOBf"
            >
              test2
            </span>
          </styled.span>
          <styled.span
            textColor="green"
            textFontFamily="Titillium Web"
            textLineHeight="1"
            textSize="15px"
            textWeight="500"
          >
            <span
              className="sc-EHOje cFtBxI"
            >
              Online
            </span>
          </styled.span>
        </div>
      </styled.div>
      <ModalDeleteChat
        currentChatId={1}
        deleteChat={[MockFunction]}
        userName="test2"
      >
        <Modal
          centered={true}
          closeOnDimmerClick={true}
          closeOnDocumentClick={false}
          dimmer={true}
          eventPool="Modal"
          onClose={[Function]}
          open={false}
          trigger={
            <Styled(Button)
              buttonheight="35px"
              buttonmargin="0 10px 0 auto"
              buttonwidth="40px"
              color="red"
              icon="user delete"
              onClick={[Function]}
            />
          }
        >
          <Portal
            closeOnDocumentClick={false}
            closeOnEscape={true}
            eventPool="Modal"
            mountNode={<body />}
            onClose={[Function]}
            onMount={[Function]}
            onOpen={[Function]}
            onUnmount={[Function]}
            open={false}
            openOnTriggerClick={true}
            trigger={
              <Styled(Button)
                buttonheight="35px"
                buttonmargin="0 10px 0 auto"
                buttonwidth="40px"
                color="red"
                icon="user delete"
                onClick={[Function]}
              />
            }
          >
            <Ref
              innerRef={[Function]}
              key="trigger"
            >
              <Styled(Button)
                buttonheight="35px"
                buttonmargin="0 10px 0 auto"
                buttonwidth="40px"
                color="red"
                icon="user delete"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                <Button
                  as="button"
                  buttonheight="35px"
                  buttonmargin="0 10px 0 auto"
                  buttonwidth="40px"
                  className="sc-dnqmqq gjfcAb"
                  color="red"
                  icon="user delete"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  role="button"
                >
                  <button
                    buttonheight="35px"
                    buttonmargin="0 10px 0 auto"
                    buttonwidth="40px"
                    className="ui red icon button sc-dnqmqq gjfcAb"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    role="button"
                  >
                    <Icon
                      as="i"
                      name="user delete"
                    >
                      <i
                        aria-hidden="true"
                        className="user delete icon"
                      />
                    </Icon>
                  </button>
                </Button>
              </Styled(Button)>
            </Ref>
          </Portal>
        </Modal>
      </ModalDeleteChat>
    </section>
  </styled.section>
</SectionContactHeader>
`;
